File: src/display/buffer.rs
//! Deterministic in-memory display used by state-machine and daemon tests.

use std::collections::VecDeque;
use std::time::Duration;

use super::{
    Dimensions, DisplayBackend, DisplayError, DisplayState, InputEvent, RestoreMode, Scene, Style,
    validate_sensitive_text,
};

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum BufferOperation {
    Acquire,
    Show,
    Hide,
    Render,
    RenderSensitiveText { row: u16, column: u16, cells: usize },
    PollInput(Duration),
    Details(bool),
    Restore(RestoreMode),
}

#[derive(Debug)]
pub struct BufferBackend {
    dimensions: Dimensions,
    state: DisplayState,
    frames: Vec<Scene>,
    input: VecDeque<InputEvent>,
    operations: Vec<BufferOperation>,
}

impl BufferBackend {
    pub fn new(dimensions: Dimensions) -> Self {
        Self {
            dimensions,
            state: DisplayState::Unacquired,
            frames: Vec::new(),
            input: VecDeque::new(),
            operations: Vec::new(),
        }
    }

    pub fn queue_input(&mut self, event: InputEvent) {
        self.input.push_back(event);
    }

    pub fn frames(&self) -> &[Scene] {
        &self.frames
    }

    pub fn operations(&self) -> &[BufferOperation] {
        &self.operations
    }

    fn require_owned(&self, operation: &'static str) -> Result<(), DisplayError> {
        if self.state.owns_resources() && self.state != DisplayState::Acquiring {
            Ok(())
        } else {
            Err(DisplayError::InvalidState {
                operation,
                state: self.state,
            })
        }
    }
}

impl DisplayBackend for BufferBackend {
    fn state(&self) -> DisplayState {
        self.state
    }

    fn dimensions(&self) -> Option<Dimensions> {
        self.state.owns_resources().then_some(self.dimensions)
    }

    fn acquire(&mut self) -> Result<(), DisplayError> {
        match self.state {
            DisplayState::Unacquired => {
                self.operations.push(BufferOperation::Acquire);
                self.state = DisplayState::Hidden;
                Ok(())
            }
            DisplayState::Hidden | DisplayState::Splash | DisplayState::Details => Ok(()),
            state => Err(DisplayError::InvalidState {
                operation: "acquire",
                state,
            }),
        }
    }

    fn show(&mut self) -> Result<(), DisplayError> {
        self.require_owned("show")?;
        if self.state != DisplayState::Splash {
            self.operations.push(BufferOperation::Show);
            self.state = DisplayState::Splash;
        }
        Ok(())
    }

    fn hide(&mut self) -> Result<(), DisplayError> {
        self.require_owned("hide")?;
        if self.state != DisplayState::Hidden {
            self.operations.push(BufferOperation::Hide);
            self.state = DisplayState::Hidden;
        }
        Ok(())
    }

    fn render(&mut self, scene: &Scene) -> Result<(), DisplayError> {
        if self.state != DisplayState::Splash {
            return Err(DisplayError::InvalidState {
                operation: "render",
                state: self.state,
            });
        }
        if scene.dimensions() != self.dimensions {
            return Err(DisplayError::SizeMismatch {
                expected: self.dimensions,
                actual: scene.dimensions(),
            });
        }
        self.operations.push(BufferOperation::Render);
        self.frames.push(scene.clone());
        Ok(())
    }

    fn render_sensitive_text(
        &mut self,
        row: u16,
        column: u16,
        text: &str,
        _style: Style,
    ) -> Result<(), DisplayError> {
        if self.state != DisplayState::Splash {
            return Err(DisplayError::InvalidState {
                operation: "render sensitive text",
                state: self.state,
            });
        }
        let cells = validate_sensitive_text(self.dimensions, row, column, text)?;
        self.operations
            .push(BufferOperation::RenderSensitiveText { row, column, cells });
        Ok(())
    }

    fn poll_input(&mut self, timeout: Duration) -> Result<Option<InputEvent>, DisplayError> {
        self.require_owned("poll input")?;
        self.operations.push(BufferOperation::PollInput(timeout));
        let event = match self.state {
            DisplayState::Splash => self.input.pop_front(),
            DisplayState::Details
                if matches!(self.input.front(), Some(InputEvent::ReturnToSplash)) =>
            {
                self.input.pop_front()
            }
            _ => None,
        };
        if let Some(InputEvent::Resized(dimensions)) = event.as_ref() {
            self.dimensions = *dimensions;
        }
        Ok(event)
    }

    fn details(&mut self, visible: bool) -> Result<(), DisplayError> {
        self.require_owned("change details visibility")?;
        let target = if visible {
            DisplayState::Details
        } else {
            DisplayState::Splash
        };
        if self.state != target {
            self.operations.push(BufferOperation::Details(visible));
            self.state = target;
        }
        Ok(())
    }

    fn restore(&mut self) -> Result<(), DisplayError> {
        self.restore_with_mode(RestoreMode::Clear)
    }

    fn restore_with_mode(&mut self, mode: RestoreMode) -> Result<(), DisplayError> {
        match self.state {
            DisplayState::Restored | DisplayState::FailedOpen => Ok(()),
            _ => {
                self.operations.push(BufferOperation::Restore(mode));
                self.state = DisplayState::Restored;
                Ok(())
            }
        }
    }
}

File: src/display/buffer.cpp
#include "sart/display/buffer.hpp"

#include <utility>

namespace sart::display {

    BufferBackend::BufferBackend(Dimensions dimensions) : dimensions_(dimensions) {}
    void BufferBackend::queue_input(InputEvent event) { input_.push_back(std::move(event)); }
    const std::vector<Scene> &BufferBackend::frames() const noexcept { return frames_; }
    const std::vector<BufferOperation> &BufferBackend::operations() const noexcept { return operations_; }
    DisplayState BufferBackend::state() const noexcept { return state_; }
    std::optional<Dimensions> BufferBackend::dimensions() const noexcept {
        return owns_resources(state_) ? std::optional{dimensions_} : std::nullopt;
    }

    void BufferBackend::require_owned(std::string_view operation) const {
        if (!owns_resources(state_) || state_ == DisplayState::acquiring) {
            throw DisplayError(DisplayErrorCode::invalid_state,
                               std::string("cannot ") + std::string(operation) + " display");
        }
    }

    void BufferBackend::acquire() {
        if (state_ == DisplayState::unacquired) {
            operations_.emplace_back(BufferOperation::Kind::acquire);
            state_ = DisplayState::hidden;
        } else if (!owns_resources(state_) || state_ == DisplayState::acquiring) {
            throw DisplayError(DisplayErrorCode::invalid_state, "cannot acquire display");
        }
    }

    void BufferBackend::show() {
        require_owned("show");
        if (state_ != DisplayState::splash) {
            operations_.emplace_back(BufferOperation::Kind::show);
            state_ = DisplayState::splash;
        }
    }

    void BufferBackend::hide() {
        require_owned("hide");
        if (state_ != DisplayState::hidden) {
            operations_.emplace_back(BufferOperation::Kind::hide);
            state_ = DisplayState::hidden;
        }
    }

    void BufferBackend::render(const Scene &scene) {
        if (state_ != DisplayState::splash) {
            throw DisplayError(DisplayErrorCode::invalid_state, "cannot render display");
        }
        if (scene.dimensions() != dimensions_) {
            throw DisplayError(DisplayErrorCode::size_mismatch, "frame size does not match display");
        }
        operations_.emplace_back(BufferOperation::Kind::render);
        frames_.push_back(scene);
    }

    void BufferBackend::render_sensitive_text(std::uint16_t row, std::uint16_t column, std::string_view text, Style) {
        if (state_ != DisplayState::splash) {
            throw DisplayError(DisplayErrorCode::invalid_state, "cannot render sensitive text");
        }
        const auto cells = validate_sensitive_text(dimensions_, row, column, text);
        BufferOperation operation(BufferOperation::Kind::render_sensitive_text);
        operation.row = row;
        operation.column = column;
        operation.cells = cells;
        operations_.push_back(operation);
    }

    std::optional<InputEvent> BufferBackend::poll_input(std::chrono::milliseconds timeout) {
        require_owned("poll input");
        BufferOperation operation{BufferOperation::Kind::poll_input};
        operation.timeout = timeout;
        operations_.push_back(operation);
        if (input_.empty())
            return std::nullopt;
        if (state_ != DisplayState::splash &&
            !(state_ == DisplayState::details && input_.front().kind() == InputEvent::Kind::return_to_splash)) {
            return std::nullopt;
        }
        auto event = std::move(input_.front());
        input_.pop_front();
        if (event.kind() == InputEvent::Kind::resized)
            dimensions_ = *event.resized_dimensions();
        return event;
    }

    void BufferBackend::details(bool visible) {
        require_owned("change details visibility");
        const auto target = visible ? DisplayState::details : DisplayState::splash;
        if (state_ != target) {
            BufferOperation operation{BufferOperation::Kind::details};
            operation.visible = visible;
            operations_.push_back(operation);
            state_ = target;
        }
    }

    void BufferBackend::restore() { restore(RestoreMode::clear); }

    void BufferBackend::restore(RestoreMode mode) {
        if (state_ == DisplayState::restored || state_ == DisplayState::failed_open)
            return;
        BufferOperation operation{BufferOperation::Kind::restore};
        operation.restore_mode = mode;
        operations_.push_back(operation);
        state_ = DisplayState::restored;
    }

} // namespace sart::display
